### PR TITLE
Ignore errors when setting keepalive period is not supported by the system

### LINKED ISSUE #7409: TCP listener crashes on OpenBSD
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

1. Starting traefik on OpenBSD (see configurations below)
2. Trying to access the published services (webservers) through traefik IP

### What did you expect to see?

The pages served by the webservers.

### What did you see instead?

Short answer:
No access: the TCP listener crashed and never recovered. Thus traefik is not useable.

Long answer:
1. Traefik is starting OK and binds OK:
```
> netstat -nap tcp |grep LISTEN | grep ".80"
tcp          0      0  *.80                   *.*                    LISTEN
```

2. First curl timeouts (the listener probably crashed and never closed the connection):
```
> curl -vLH 'Host: my.domain'  http://127.0.0.1
*   Trying 127.0.0.1:80...
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host: 127.0.0.1
> User-Agent: curl/7.69.1
> Accept: */*
>

^C
```

3. And other curls can't even access anymore:
```
> curl -vLH 'Host: my.domain' http://127.0.0.1
*   Trying 127.0.0.1:80...
* connect to 127.0.0.1 port 80 failed: Connection refused
* Failed to connect to 127.0.0.1 port 80: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 127.0.0.1 port 80: Connection refused
```

4. Indeed, the listener failed (see logs below):
```
> netstat -nap tcp |grep LISTEN | grep ".80"
```

5. You can see the culprit in the logs:
```
time="2020-10-10T13:47:58Z" level=error msg="set tcp 127.0.0.1:80->127.0.0.1:34520: protocol not available" entryPointName=web
time="2020-10-10T13:47:58Z" level=error msg="Error while starting server: set tcp 127.0.0.1:80->127.0.0.1:34520: protocol not available" entryPointName=web
time="2020-10-10T13:47:58Z" level=error msg="Error while starting server: set tcp 127.0.0.1:80->127.0.0.1:34520: protocol not available" entryPointName=web
```

6. The problem is: OpenBSD has no user-settable per-socket TCP keepalive

In traefik, the SetKeepAlivePeriod function is called for new TCP connections and an error is returned if necessary: https://github.com/traefik/traefik/blob/5c853766e8cd0328e9c9393d6d99503d778f889b/pkg/server/server_entrypoint_tcp.go#L342

In go sources, we can see that OpenBSD always returns an error (ENOPROTOOPT):
https://github.com/golang/go/blob/dev.boringcrypto.go1.15/src/net/tcpsockopt_openbsd.go

=> pr in progress

Other projects had the bug too:
https://github.com/labstack/echo/pull/1456
https://github.com/aerogo/aero/issues/10
https://github.com/caddyserver/caddy/pull/2787

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
> traefik version
Version:      2.3.1
Codename:     picodon
Go version:   go1.15.2
Built:        2020-09-29T15:49:06Z
OS/Arch:      openbsd/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

OpenBSD version:
```
OpenBSD 6.7 GENERIC#179 amd64
```

Traefik configuration:
```yaml
entryPoints:
  web:
    address: ":80"

providers:
  consulCatalog:
    prefix: "traefik"
    exposedByDefault: false

    endpoint:
      address: "http://127.0.0.1:8500"

log:
  level: "DEBUG"
```

Consul service configuration:
```
      service {
        name = "www"
        port = "http"

        tags = [
          "traefik.enable=true",
          "traefik.http.routers.www.rule=Host(`my.domain`)",
          "traefik.http.routers.www.entrypoints=web"
        ]

        check {
          type = "http"
          port = "http"
          path = "/"
          interval = "30s"
          timeout = "5s"
        }
      }
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
time="2020-10-10T13:47:36Z" level=info msg="Traefik version 2.3.1 built on 2020-09-29T15:49:06Z"
time="2020-10-10T13:47:36Z" level=debug msg="Static configuration loaded {\"global\":{\"checkNewVersion\":true},\"serversTransport\":{\"maxIdleConnsPerHost\":200},\"entryPoints\":{\"web\":{\"address\":\":80\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{},\"http\":{}}},\"providers\":{\"providersThrottleDuration\":2000000000,\"consulCatalog\":{\"endpoint\":{\"address\":\"http://127.0.0.1:8500\"},\"prefix\":\"traefik\",\"refreshInterval\":15000000000,\"defaultRule\":\"Host(`{{ normalize .Name }}`)\"}},\"log\":{\"level\":\"DEBUG\",\"format\":\"common\"}}"
time="2020-10-10T13:47:36Z" level=info msg="\nStats collection is disabled.\nHelp us improve Traefik by turning this feature on :)\nMore details on: https://doc.traefik.io/traefik/contributing/data-collection/\n"
time="2020-10-10T13:47:36Z" level=info msg="Starting provider aggregator.ProviderAggregator {}"
time="2020-10-10T13:47:36Z" level=debug msg="Start TCP Server" entryPointName=web
time="2020-10-10T13:47:36Z" level=info msg="Starting provider *traefik.Provider {}"
time="2020-10-10T13:47:36Z" level=debug msg="Configuration received from provider internal: {\"http\":{\"services\":{\"noop\":{}}},\"tcp\":{},\"tls\":{}}" providerName=internal
time="2020-10-10T13:47:36Z" level=info msg="Starting provider *consulcatalog.Provider {\"endpoint\":{\"address\":\"http://127.0.0.1:8500\"},\"prefix\":\"traefik\",\"refreshInterval\":15000000000,\"defaultRule\":\"Host(`{{ normalize .Name }}`)\"}"
time="2020-10-10T13:47:36Z" level=debug msg="No default certificate, generating one"
time="2020-10-10T13:47:51Z" level=debug msg="Filtering disabled item" serviceName=traefik
time="2020-10-10T13:47:51Z" level=debug msg="Filtering disabled item" serviceName=consul
time="2020-10-10T13:47:51Z" level=debug msg="Filtering disabled item" serviceName=nomad
time="2020-10-10T13:47:51Z" level=debug msg="Filtering disabled item" serviceName=nomad-client
time="2020-10-10T13:47:51Z" level=debug msg="Configuration received from provider consulcatalog: {\"http\":{\"routers\":{\"www\":{\"entryPoints\":[\"web\"],\"service\":\"www\",\"rule\":\"Host(`my.domain`)\"}},\"services\":{\"www\":{\"loadBalancer\":{\"servers\":[{\"url\":\"http://10.20.128.229:22042\"}],\"passHostHeader\":true}}}},\"tcp\":{},\"udp\":{}}" providerName=consulcatalog
time="2020-10-10T13:47:51Z" level=debug msg="Creating middleware" entryPointName=web routerName=www@consulcatalog serviceName=www middlewareName=pipelining middlewareType=Pipelining
time="2020-10-10T13:47:51Z" level=debug msg="Creating load-balancer" entryPointName=web routerName=www@consulcatalog serviceName=www
time="2020-10-10T13:47:51Z" level=debug msg="Creating server 0 http://10.20.128.229:22042" serviceName=www entryPointName=web routerName=www@consulcatalog serverName=0
time="2020-10-10T13:47:51Z" level=debug msg="Added outgoing tracing middleware www" routerName=www@consulcatalog middlewareName=tracing middlewareType=TracingForwarder entryPointName=web
time="2020-10-10T13:47:51Z" level=debug msg="Creating middleware" middlewareName=traefik-internal-recovery entryPointName=web middlewareType=Recovery
time="2020-10-10T13:47:51Z" level=debug msg="No default certificate, generating one"
time="2020-10-10T13:47:58Z" level=error msg="set tcp 127.0.0.1:80->127.0.0.1:34520: protocol not available" entryPointName=web
time="2020-10-10T13:47:58Z" level=error msg="Error while starting server: set tcp 127.0.0.1:80->127.0.0.1:34520: protocol not available" entryPointName=web
time="2020-10-10T13:47:58Z" level=error msg="Error while starting server: set tcp 127.0.0.1:80->127.0.0.1:34520: protocol not available" entryPointName=web
time="2020-10-10T13:48:06Z" level=debug msg="Filtering disabled item" serviceName=consul
time="2020-10-10T13:48:06Z" level=debug msg="Filtering disabled item" serviceName=nomad
time="2020-10-10T13:48:06Z" level=debug msg="Filtering disabled item" serviceName=nomad-client
time="2020-10-10T13:48:06Z" level=debug msg="Filtering disabled item" serviceName=traefik
time="2020-10-10T13:48:06Z" level=debug msg="Configuration received from provider consulcatalog: {\"http\":{\"routers\":{\"www\":{\"entryPoints\":[\"web\"],\"service\":\"www\",\"rule\":\"Host(`my.domain`)\"}},\"services\":{\"www\":{\"loadBalancer\":{\"servers\":[{\"url\":\"http://10.20.128.229:22042\"}],\"passHostHeader\":true}}}},\"tcp\":{},\"udp\":{}}" providerName=consulcatalog
```

### PR DESCRIPTION
**Pull request by tristan-weil**
### What does this PR do?

Fix #7409 

### More

- [ ] Added/updated tests
- [ ] Added/updated documentation

### Additional Notes

2 others ways for the fix: 
- completely ignoring the error, if any
- making the call to the function conditionnal with a runtime.GOOS=="openbsd"

### CODE REVIEW

**Comment by rtribotte** [comment]
Hello @tristan-weil,

Thanks for your contribution !

We appreciate the way you handle the error, that allows other platforms that OpenBSD to benefits of this fix.
Can you rebase your PR on the upstream v2.3 ?

File: pkg/server/server_entrypoint_tcp.go [hunk 1]
```diff
@@ -340,7 +342,11 @@ func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
 	}
 
 	if err = tc.SetKeepAlivePeriod(3 * time.Minute); err != nil {
-		return nil, err
+		// Some systems, such as OpenBSD, have no user-settable per-socket TCP
+		// keepalive options
```
Comment (jspdown) [inline pkg/server/server_entrypoint_tcp.go:346]:
```suggestion
		// keepalive options.
```

**Comment by tristan-weil** [comment]
rebase done.

thanks!

**Review by jspdown — APPROVED** [review]
LGTM :+1:

**Review by rtribotte — APPROVED** [review]
LGTM 👍

**Review by jbdoumenjou — APPROVED** [review]
LGTM

**Review by kevinpollet — APPROVED** [review]
LGTM 🚢

**Pull request merged**